elephant/spike_train_dissimilarity.py

```diff
@@ -0,0 +1,406 @@
+# -*- coding: utf-8 -*-
+"""
+In neuroscience one often wants to evaluate, how similar or dissimilar pairs
+or even large sets of spiketrains are. For this purpose various different
+spike train dissimilarity measures were introduced in the literature.
+They differ, e.g., by the properties of having the mathematical properties of
+a metric or by being time-scale dependent or not. Well known representatives
+of spike train dissimilarity measures are the Victor-Purpura distance and the
+Van Rossum distance implemented in this module, which both are metrics in the
+mathematical sense and time-scale dependent.
+
+:copyright: Copyright 2015 by the Elephant team, see AUTHORS.txt.
+:license: Modified BSD, see LICENSE.txt for details.
+"""
+
+import quantities as pq
+import numpy as np
+import scipy as sp
+import elephant.kernels as kernels
+import cmath
+from neo.core import SpikeTrain
+
+# Problem of conversion from Python 2 to Python 3:
+# 'xrange' in Python 2 is 'range' in Python 3. 
+try:
+    xrange
+except NameError:
+    xrange = range
+
+
+def _create_matrix_from_indexed_function(
+        shape, func, symmetric_2d=False, **func_params):
+    mat = np.empty(shape)
+    if symmetric_2d:
+        for i in xrange(shape[0]):
+            for j in xrange(i, shape[1]):
+                mat[i, j] = mat[j, i] = func(i, j, **func_params)
+    else:
+        for idx in np.ndindex(*shape):
+            mat[idx] = func(*idx, **func_params)
+    return mat
+
+
+def victor_purpura_dist(
+        trains, q=1.0 * pq.Hz, kernel=None, sort=True, algor='fast_algor'):
+    """ 
+    Calculates the Victor-Purpura's (VP) distance. It is often denoted as
+    :math:`D^{\\text{spike}}[q]`.
+
+    It is defined as the minimal cost of transforming spike train `a` into
+    spike train `b` by using the following operations:
+
+        * Inserting or deleting a spike (cost 1.0).
+        * Shifting a spike from :math:`t` to :math:`t'` (cost :math:`q 
+          \\cdot |t - t'|`).
+
+    A detailed description can be found in
+    *Victor, J. D., & Purpura, K. P. (1996). Nature and precision of
+    temporal coding in visual cortex: a metric-space analysis. Journal of
+    Neurophysiology.*
+
+    Given the average number of spikes :math:`n` in a spike train and
+    :math:`N` spike trains the run-time complexity of this function is
+    :math:`O(N^2 n^2)` and :math:`O(N^2 + n^2)` memory will be needed.
+
+    Parameters
+    ----------
+    trains : Sequence of :class:`neo.core.SpikeTrain` objects of
+        which the distance will be calculated pairwise.
+    q: Quantity scalar 
```

Comment:
I see that below you explicitly allow for q=0 and q=np.inf. Please indicate these two special cases and their meaning here (otherwise users will not know that they can actually specify inifinite kernels).